File: services/messaging-gateway/internal/channel/ussd.go
// USSD channel support (SPEC-W12 Agent A): session state (180s TTL), the
// pack-driven numeric menu model, the tenant menu fetch and the synchronous
// request/reply client to conversation-service.
//
// Contract (SPEC-W12 §1): Africa's Talking posts form fields sessionId,
// serviceCode, phoneNumber, text (text is the CUMULATIVE "1*2*3" input,
// empty on the first request of a session). The gateway answers text/plain
// prefixed "CON " (session continues) or "END " (session terminates).
//
// Request/reply contract with conversation-service (DELIVERED by Agent D,
// app/ussd.py + routes.py): every callback is forwarded synchronously via
// the same Dapr invoke path the other inbound channels use
// (http://127.0.0.1:{DAPR_HTTP_PORT}/v1.0/invoke/conversation-service/method,
// overridable with CONVERSATION_URL) as POST {base}/v1/ussd/turns with the
// tenant pack's resolved ussd.menu attached when defined (menu mode is
// driven conversation-side: numeric selection, 0=back, 00=main menu).
// conversation-service returns {conversation_id, reply, continue, mode,
// selection, action} in the invoke response body; the gateway renders
// "CON reply" when continue=true, "END reply" otherwise.
package channel

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"sync"
	"time"
)

// USSDSessionTTL is the session state TTL (SPEC-W12 §1: 180s).
const USSDSessionTTL = 180 * time.Second

// USSDMenuItem is one entry of a tenant pack's ussd.menu list
// ({key,label,action}).
type USSDMenuItem struct {
	Key    string `json:"key"`
	Label  string `json:"label"`
	Action string `json:"action"`
}

// ParseUSSDMenu extracts pack.ussd.menu from an identity tenant payload
// (GET /v1/tenants/{slug} → {"pack": {...}}).
//
// ASSUMPTION: identity's pack Summary passes the pack yaml `ussd:` block
// through as JSON {"ussd": {"menu": [...]}}. That passthrough does not exist
// in identity-service yet (flagged for a follow-up), so today every tenant
// resolves to an empty menu and the gateway runs in pass-through text mode —
// the SPEC-mandated fallback. Items without key/label are skipped.
func ParseUSSDMenu(tenantPayload map[string]any) []USSDMenuItem {
	pack, ok := tenantPayload["pack"].(map[string]any)
	if !ok {
		return nil
	}
	ussd, ok := pack["ussd"].(map[string]any)
	if !ok {
		return nil
	}
	raw, ok := ussd["menu"].([]any)
	if !ok {
		return nil
	}
	menu := make([]USSDMenuItem, 0, len(raw))
	for _, entry := range raw {
		m, ok := entry.(map[string]any)
		if !ok {
			continue
		}
		item := USSDMenuItem{
			Key:    asString(m["key"]),
			Label:  asString(m["label"]),
			Action: asString(m["action"]),
		}
		if item.Key == "" || item.Label == "" {
			continue
		}
		menu = append(menu, item)
	}
	return menu
}

func asString(v any) string {
	s, _ := v.(string)
	return s
}

// NOTE: menu rendering and the numeric navigation state machine
// (1/2/… select, 0=back, 00=main menu) live conversation-side in Agent D's
// app/ussd.py — the gateway resolves the pack menu once per session and
// attaches it to every turn (USSDTurnRequest.Menu).

// ---------------------------------------------------------------------------
// Session state
// ---------------------------------------------------------------------------

// USSDSession is the per-sessionId state kept between callbacks.
type USSDSession struct {
	ID          string         `json:"id"`           // provider sessionId
	ServiceCode string         `json:"service_code"` // e.g. *384*123#
	Phone       string         `json:"phone"`        // E.164
	SiteSlug    string         `json:"site_slug"`    // CHANNEL_SITE_MAP resolution
	TenantID    string         `json:"tenant_id"`
	Menu        []USSDMenuItem `json:"menu,omitempty"` // resolved pack menu; nil = pass-through text mode
	UpdatedAt   time.Time      `json:"updated_at"`
}

func (s *USSDSession) clone() *USSDSession {
	out := *s
	if s.Menu != nil {
		out.Menu = make([]USSDMenuItem, len(s.Menu))
		copy(out.Menu, s.Menu)
	}
	return &out
}

// USSDSessionStore keeps session state for the callback round-trips.
type USSDSessionStore interface {
	// Get returns the session, or (nil, nil) when absent or expired.
	Get(ctx context.Context, id string) (*USSDSession, error)
	Save(ctx context.Context, s *USSDSession, ttl time.Duration) error
	Delete(ctx context.Context, id string) error
}

// MemoryUSSDStore is the default store: in-process map with lazy TTL
// expiry. The messaging-gateway is deployed as a single replica per
// environment (see internal/metrics), so process-local state matches the
// established scaling model; USSD_SESSION_BACKEND=dapr switches to the Dapr
// state store for multi-replica deployments.
type MemoryUSSDStore struct {
	mu  sync.Mutex
	m   map[string]memoryUSSDEntry
	now func() time.Time // injectable for tests
}

type memoryUSSDEntry struct {
	sess      *USSDSession
	expiresAt time.Time
}

// NewMemoryUSSDStore builds an empty in-memory store.
func NewMemoryUSSDStore() *MemoryUSSDStore {
	return &MemoryUSSDStore{m: map[string]memoryUSSDEntry{}, now: time.Now}
}

// Get returns the session or (nil, nil) when missing/expired.
func (s *MemoryUSSDStore) Get(_ context.Context, id string) (*USSDSession, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	e, ok := s.m[id]
	if !ok {
		return nil, nil
	}
	if s.now().After(e.expiresAt) {
		delete(s.m, id)
		return nil, nil
	}
	return e.sess.clone(), nil
}

// Save upserts the session with a fresh TTL.
func (s *MemoryUSSDStore) Save(_ context.Context, sess *USSDSession, ttl time.Duration) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.m[sess.ID] = memoryUSSDEntry{sess: sess.clone(), expiresAt: s.now().Add(ttl)}
	return nil
}

// Delete drops the session (session end / explicit reset).
func (s *MemoryUSSDStore) Delete(_ context.Context, id string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.m, id)
	return nil
}

// DaprUSSDStore keeps sessions in a Dapr state store (mirrors the daprc
// net/http state pattern used by booking/identity/notification — no Dapr
// SDK). The TTL is delegated to the state store via the ttlInSeconds
// metadata key.
//
// INFRA NOTE: infra/dapr/components/statestore.redis.yaml does not list the
// messaging-gateway app-id in its scopes today (flagged for Agent D); the
// store name is configurable via USSD_STATE_STORE.
type DaprUSSDStore struct {
	base  string // http://127.0.0.1:{DAPR_HTTP_PORT}
	store string // state store component name
	hc    *http.Client
}

// NewDaprUSSDStore builds the store client; store defaults to "statestore".
func NewDaprUSSDStore(store string, daprHTTPPort int) *DaprUSSDStore {
	if store == "" {
		store = "statestore"
	}
	return &DaprUSSDStore{
		base:  fmt.Sprintf("http://127.0.0.1:%d", daprHTTPPort),
		store: store,
		hc:    &http.Client{Timeout: 5 * time.Second},
	}
}

func (s *DaprUSSDStore) key(id string) string { return "ussd:" + id }

// Get returns the session or (nil, nil) when missing/expired (the state
// store enforces the TTL and answers 204 No Content).
func (s *DaprUSSDStore) Get(ctx context.Context, id string) (*USSDSession, error) {
	u := fmt.Sprintf("%s/v1.0/state/%s/%s", s.base, s.store, s.key(id))
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.hc.Do(req)
	if err != nil {
		return nil, fmt.Errorf("state get %s: %w", s.key(id), err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusNoContent || resp.StatusCode == http.StatusNotFound {
		return nil, nil
	}
	if resp.StatusCode >= 300 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return nil, fmt.Errorf("state get %s: status %d: %s", s.key(id), resp.StatusCode, b)
	}
	var sess USSDSession
	if err := json.NewDecoder(resp.Body).Decode(&sess); err != nil {
		return nil, fmt.Errorf("state get %s: decode: %w", s.key(id), err)
	}
	return &sess, nil
}

// Save upserts the session with the store-side TTL.
func (s *DaprUSSDStore) Save(ctx context.Context, sess *USSDSession, ttl time.Duration) error {
	body, err := json.Marshal([]map[string]any{{
		"key":      s.key(sess.ID),
		"value":    sess,
		"metadata": map[string]string{"ttlInSeconds": strconv.Itoa(int(ttl.Seconds()))},
	}})
	if err != nil {
		return err
	}
	u := fmt.Sprintf("%s/v1.0/state/%s", s.base, s.store)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := s.hc.Do(req)
	if err != nil {
		return fmt.Errorf("state save %s: %w", s.key(sess.ID), err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return fmt.Errorf("state save %s: status %d: %s", s.key(sess.ID), resp.StatusCode, b)
	}
	return nil
}

// Delete drops the session.
func (s *DaprUSSDStore) Delete(ctx context.Context, id string) error {
	u := fmt.Sprintf("%s/v1.0/state/%s/%s", s.base, s.store, s.key(id))
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, u, nil)
	if err != nil {
		return err
	}
	resp, err := s.hc.Do(req)
	if err != nil {
		return fmt.Errorf("state delete %s: %w", s.key(id), err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 && resp.StatusCode != http.StatusNotFound {
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return fmt.Errorf("state delete %s: status %d: %s", s.key(id), resp.StatusCode, b)
	}
	return nil
}

// ---------------------------------------------------------------------------
// Upstream clients
// ---------------------------------------------------------------------------

// ResolveInvokeBase maps a direct-base override onto the Dapr sidecar
// invoke base for an app-id (same convention as ResolveBases).
func ResolveInvokeBase(override, appID string, daprHTTPPort int) string {
	if override != "" {
		return override
	}
	return fmt.Sprintf("http://127.0.0.1:%d/v1.0/invoke/%s/method", daprHTTPPort, appID)
}

// USSDMenuFetcher loads the tenant pack's ussd.menu (nil = pass-through
// text mode). Implemented by HTTPUSSDMenuFetcher in prod, faked in tests.
type USSDMenuFetcher interface {
	USSDMenu(ctx context.Context, tenantSlug string) ([]USSDMenuItem, error)
}

// HTTPUSSDMenuFetcher fetches the menu via the identity packs summary
// endpoint (GET {base}/v1/tenants/{slug}, same client pattern as the voice
// runtime's tenant-context fetch — Dapr invoke app-id identity, overridable
// with IDENTITY_URL). tenantSlug is the CHANNEL_SITE_MAP site_slug for the
// USSD service code (ASSUMPTION: for USSD deployments the site slug is the
// tenant slug — single-site tenants, the common case for the NG channel).
type HTTPUSSDMenuFetcher struct {
	Base string
	HC   *http.Client
}

// NewUSSDMenuFetcher builds the fetcher with a 5s-timeout client.
func NewUSSDMenuFetcher(identityBase string) *HTTPUSSDMenuFetcher {
	return &HTTPUSSDMenuFetcher{Base: identityBase, HC: &http.Client{Timeout: 5 * time.Second}}
}

// USSDMenu returns the parsed menu, or nil when the pack defines none.
func (f *HTTPUSSDMenuFetcher) USSDMenu(ctx context.Context, tenantSlug string) ([]USSDMenuItem, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet,
		f.Base+"/v1/tenants/"+tenantSlug, nil)
	if err != nil {
		return nil, err
	}
	resp, err := f.HC.Do(req)
	if err != nil {
		return nil, fmt.Errorf("identity tenant fetch: %w", err)
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 256*1024))
	if resp.StatusCode >= 300 {
		return nil, fmt.Errorf("identity tenant fetch: status %d", resp.StatusCode)
	}
	var payload map[string]any
	if err := json.Unmarshal(raw, &payload); err != nil {
		return nil, fmt.Errorf("identity tenant fetch: decode: %w", err)
	}
	return ParseUSSDMenu(payload), nil
}

// USSDTurnRequest is the synchronous request/reply contract with
// conversation-service (POST {base}/v1/ussd/turns — DELIVERED by Agent D,
// app/ussd.py UssdTurnRequest; keep the field set exactly in sync).
type USSDTurnRequest struct {
	TenantID    string `json:"tenant_id"`  // uuid — conversation key is uuid5(tenant, sessionId)
	SiteSlug    string `json:"site_slug"`  // required by conversation-service
	SessionID   string `json:"session_id"` // provider sessionId
	ServiceCode string `json:"service_code,omitempty"`
	PhoneNumber string `json:"phone_number,omitempty"`
	Text        string `json:"text"` // cumulative "1*2*3" input as received ("" on first request)
	// Menu is the tenant pack's resolved ussd.menu; its presence switches
	// the session to menu mode conversation-side. Omitted = pass-through
	// text mode.
	Menu []USSDMenuItem `json:"menu,omitempty"`
}

// USSDTurnResponse is what conversation-service returns in the invoke
// response body (Agent D response_payload); the gateway renders
// "CON reply" when Continue is true, "END reply" otherwise.
type USSDTurnResponse struct {
	ConversationID string `json:"conversation_id"`
	Reply          string `json:"reply"`
	Continue       bool   `json:"continue"` // true → session stays open (CON)
	Mode           string `json:"mode"`     // "menu" | "text"
	Selection      string `json:"selection"`
	Action         string `json:"action"`
}

// USSDConversation is the conversation-service USSD turn contract
// (implemented by HTTPUSSDConversation in prod, faked in tests).
type USSDConversation interface {
	Turn(ctx context.Context, req USSDTurnRequest) (USSDTurnResponse, error)
}

// HTTPUSSDConversation invokes conversation-service via the same Dapr
// invoke path the other inbound channels use (CONVERSATION_URL override or
// {DAPR}/v1.0/invoke/conversation-service/method).
type HTTPUSSDConversation struct {
	Base string
	HC   *http.Client
}

// NewUSSDConversation builds the client with a 10s-timeout HTTP client
// (matches the bridge's client).
func NewUSSDConversation(conversationBase string) *HTTPUSSDConversation {
	return &HTTPUSSDConversation{Base: conversationBase, HC: &http.Client{Timeout: 10 * time.Second}}
}

// Turn posts one USSD interaction and decodes the reply.
func (c *HTTPUSSDConversation) Turn(ctx context.Context, turn USSDTurnRequest) (USSDTurnResponse, error) {
	var out USSDTurnResponse
	body, err := json.Marshal(turn)
	if err != nil {
		return out, err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.Base+"/v1/ussd/turns", bytes.NewReader(body))
	if err != nil {
		return out, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.HC.Do(req)
	if err != nil {
		return out, fmt.Errorf("conversation ussd turn: %w", err)
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 256*1024))
	if resp.StatusCode >= 300 {
		return out, fmt.Errorf("conversation ussd turn: status %d: %s", resp.StatusCode, truncateBody(raw))
	}
	if err := json.Unmarshal(raw, &out); err != nil {
		return out, fmt.Errorf("conversation ussd turn: decode: %w", err)
	}
	return out, nil
}

func truncateBody(b []byte) string {
	if len(b) > 512 {
		return string(b[:512])
	}
	return string(b)
}

File: services/conversation-service/app/ussd.py
"""USSD session → conversation mapping (SPEC-W12, cross-agent contract §1/§2).

messaging-gateway terminates the Africa's Talking USSD callback (form fields
``sessionId, serviceCode, phoneNumber, text`` — ``text`` is the CUMULATIVE
``1*2*3`` input, empty on the first request of a session) and invokes this
service synchronously via Dapr (``POST /v1/ussd/turns``). This module owns:

- the deterministic session → conversation key: ``uuid5(tenant, sessionId)``
  so every callback of one USSD session lands in the SAME conversation and
  session resumption needs no lookup table (no extra state store — the
  180s session TTL from contract §1 is enforced by messaging-gateway's
  session store);
- low-literacy MENU MODE: when the tenant pack defines ``ussd.menu`` the
  gateway passes the resolved menu (list of ``{key,label,action}``) in the
  invoke body and this module drives the numeric navigation —
  ``1``/``2``/… select an item, ``0`` = back, ``00`` = main menu;
- the request/reply contract: the user turn text is the cumulative input
  parsed down to the LAST selection; the invoke response body carries the
  reply text plus a ``continue`` flag, and messaging-gateway renders the
  wire format (``CON <reply>`` when continue=true, ``END <reply>`` when
  false, text/plain per contract §1).

Channel enum: the conversation is persisted with ``channel="ussd"``
(contract §2). USER turns flow through the SAME enrichment + incident
classifier path as every other channel (routes.py ``_persist_turn``); the
classifier treats ussd as web-like (incidents._CHANNEL_MAP: ussd→web).
Agent/system turns are never incident-classified (existing rule).
"""

from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, Field

CHANNEL = "ussd"

# Low-literacy navigation keys (contract with messaging-gateway + tenant
# pack authors): "0" goes back one level, "00" returns to the main menu.
# Our menus are single-level, so both re-render the main menu.
KEY_BACK = "0"
KEY_MAIN_MENU = "00"

CONFIRMATION_SUFFIX = (
    " Thank you — your request has been received and you will get a "
    "confirmation SMS shortly."
)
INVALID_SELECTION_PREFIX = "Invalid selection. Please try again.\n"


# ---------------------------------------------------------------------------
# Request/response models (Dapr invoke body contract with messaging-gateway)
# ---------------------------------------------------------------------------


class UssdMenuItem(BaseModel):
    """One tenant-pack ``ussd.menu`` entry ({key,label,action})."""

    key: str = Field(min_length=1, max_length=8)
    label: str = Field(min_length=1, max_length=160)
    action: str = ""


class UssdTurnRequest(BaseModel):
    """One Africa's Talking callback, normalized by messaging-gateway.

    ``text`` is the cumulative session input (``1*2*3``; empty on the first
    request). ``menu`` is the tenant pack's ``ussd.menu`` when defined —
    its presence switches the session to menu mode; absent = pass-through
    text mode (free text forwarded to the conversation).
    """

    tenant_id: uuid.UUID
    site_slug: str = Field(min_length=1, max_length=128)
    session_id: str = Field(min_length=1, max_length=128)
    service_code: str = Field(default="", max_length=32)
    phone_number: str | None = Field(default=None, max_length=64)
    text: str = Field(default="", max_length=182)  # AT caps USSD input ~182 chars
    menu: list[UssdMenuItem] | None = None


# ---------------------------------------------------------------------------
# Session → conversation mapping
# ---------------------------------------------------------------------------


def session_conversation_id(tenant_id: uuid.UUID, session_id: str) -> uuid.UUID:
    """Deterministic conversation key for one USSD session (contract §1).

    uuid5(tenant, sessionId) — the same callback stream always maps to the
    same conversation, for every tenant, without any lookup state.
    """
    return uuid.uuid5(
        uuid.NAMESPACE_URL, f"opendesk:ussd:{tenant_id}:{session_id}"
    )


# ---------------------------------------------------------------------------
# Cumulative-input parsing (Africa's Talking ``text`` field)
# ---------------------------------------------------------------------------


def selection_path(text: str) -> list[str]:
    """Split cumulative input into the selection path ("1*2*3" -> [1,2,3])."""
    if not text:
        return []
    return [seg.strip() for seg in text.split("*") if seg.strip()]


def parse_last_selection(text: str) -> str:
    """The LAST selection of the cumulative input ("" on the first request)."""
    path = selection_path(text)
    return path[-1] if path else ""


# ---------------------------------------------------------------------------
# Menu mode (low-literacy numeric navigation)
# ---------------------------------------------------------------------------


def render_menu(menu: list[UssdMenuItem]) -> str:
    """Numbered main-menu text with the 0/00 navigation footer."""
    lines = [f"{item.key}. {item.label}" for item in menu]
    lines.append(f"{KEY_BACK}. Back")
    lines.append(f"{KEY_MAIN_MENU}. Main menu")
    return "\n".join(lines)


def match_selection(
    menu: list[UssdMenuItem], selection: str
) -> UssdMenuItem | None:
    for item in menu:
        if item.key == selection:
            return item
    return None


def menu_reply(
    menu: list[UssdMenuItem], text: str
) -> tuple[str, bool, UssdMenuItem | None]:
    """Menu-mode state machine. Returns (reply, continue, selected_item).

    - first request (empty text), ``0`` (back), ``00`` (main menu):
      re-render the menu, session continues;
    - a key matching a menu item: confirmation text, session ENDS
      (single-level menus — the leaf action is terminal);
    - anything else: invalid-selection prompt + menu, session continues.
    """
    selection = parse_last_selection(text)
    if not selection or selection in (KEY_BACK, KEY_MAIN_MENU):
        return render_menu(menu), True, None
    item = match_selection(menu, selection)
    if item is not None:
        return f"{item.label}.{CONFIRMATION_SUFFIX}", False, item
    return INVALID_SELECTION_PREFIX + render_menu(menu), True, None


def build_reply(
    body: UssdTurnRequest, text_mode_reply: str
) -> tuple[str, bool, UssdMenuItem | None]:
    """(reply, continue, selected_item) for one callback.

    Menu mode when the pack menu was passed, else pass-through text mode
    (the configurable acknowledgement; the gateway owns session length).
    """
    if body.menu:
        return menu_reply(body.menu, body.text)
    return text_mode_reply, True, None


# ---------------------------------------------------------------------------
# User-turn text (cumulative input parsed to the LAST selection)
# ---------------------------------------------------------------------------


def user_turn_text(body: UssdTurnRequest) -> str:
    """The user-turn text persisted for one callback.

    Menu mode stores the human-readable label of the LAST selection (so the
    transcript stays meaningful for low-literacy sessions and the incident
    classifier still sees e.g. a "Report emergency" pick); text mode stores
    the raw last selection. The first request records a session-start
    marker so the conversation timeline shows how it began.
    """
    selection = parse_last_selection(body.text)
    if not selection:
        via = f" via {body.service_code}" if body.service_code else ""
        return f"(ussd session started{via})"
    if body.menu:
        if selection in (KEY_BACK, KEY_MAIN_MENU):
            return "(ussd: back to main menu)"
        item = match_selection(body.menu, selection)
        if item is not None:
            return f"{item.label} (ussd menu option {item.key})"
        return f"ussd invalid selection: {selection}"
    return selection


def idempotency_key(body: UssdTurnRequest) -> str:
    """Per-callback dedupe key: identical AT retries collapse to one turn.

    The cumulative text makes each step of a session unique; a retry of the
    SAME callback carries the same text and is deduped by the existing
    Idempotency-Key machinery in routes.py.
    """
    return f"ussd:{body.session_id}:{body.text or '<start>'}"


def response_payload(
    *,
    conversation_id: uuid.UUID,
    reply: str,
    continue_session: bool,
    body: UssdTurnRequest,
    selected: UssdMenuItem | None,
) -> dict[str, Any]:
    """Invoke response body consumed by messaging-gateway.

    Contract (docs/runbook-wave12.md + docs/channels-ussd.md): the gateway
    prefixes ``CON `` when ``continue`` is true, ``END `` otherwise, and
    answers Africa's Talking as text/plain.
    """
    return {
        "conversation_id": str(conversation_id),
        "reply": reply,
        "continue": continue_session,
        "mode": "menu" if body.menu else "text",
        "selection": parse_last_selection(body.text),
        "action": selected.action if selected else "",
    }
